### ISSUE TITLE
Attack success rate of ZOO Attack on MNIST much lower than the reported in literature

### ISSUE THREAD

**Issue by amir1m** [Trusted-AI/adversarial-robustness-toolbox#1078]
Hello,
I am trying ZOO attack from ART. However, the attack success rate (ASR) I am getting for random 100 samples from MNIST is much lower than 100% that is reported in literature. I am using following CNN same as used in literature (https://arxiv.org/abs/1805.11090 , https://arxiv.org/abs/1909.07490):

  ```
model = Sequential()

  model.add(Conv2D(32, (3, 3),activation="relu", input_shape=(28, 28, 1)))

  model.add(Conv2D(32, (3, 3), activation="relu"))
  model.add(MaxPooling2D(pool_size=(2, 2)))
  model.add(Conv2D(64, (3, 3), activation="relu"))
  model.add(Conv2D(64, (3, 3), activation="relu"))
  model.add(MaxPooling2D(pool_size=(2, 2)))
  model.add(Flatten())
  model.add(Dense(200, activation="relu"))
  model.add(Dense(200, activation="relu"))
  model.add(Dense(10, activation='softmax'))
  
  sgd = SGD(lr=0.01, decay=1e-6, momentum=0.9, nesterov=True)
  model.compile(loss="categorical_crossentropy", optimizer=sgd, metrics=["accuracy"])

```
And ZOO attack parameters used are as follows:
```
attack_cnn = ZooAttack(
    classifier=classifier,
    confidence=0.0,
    targeted=False,
    learning_rate=1e-2,
    max_iter=3000,
    binary_search_steps=10,
    initial_const=1e-3,
    abort_early=True,
    use_resize=False,
    use_importance=False,
    nb_parallel=128,
    batch_size=1,
    variable_h=0.01,
)
```

What am i missing?

Thanks,
-Amir

**Comment by beat-buesser** [comment]
Hi @amir1m

Thank you very much for using ART! It looks like your model is predicting softmax probabilities whereas the original ZOO attack is expecting logits as model output/predictions. Let me know if this improves your experiment!

**Comment by amir1m** [comment]
Thanks @beat-buesser for your reply!

I removed the `softmax` activation and used following custom loss as used in original [ZOO implementation](https://github.com/IBM/ZOO-Attack/blob/master/train_models.py)

```
def fn(correct, predicted):
        return tf.nn.softmax_cross_entropy_with_logits(labels=correct,
                                                       logits=predicted)
```
This model compiles and trains with original `model.fit()`. However, with `ART KerasClassifier()` instantiation as:

`classifier = KerasClassifier(model=model)
`

It gives error as:

```
/usr/local/lib/python3.7/dist-packages/art/estimators/classification/keras.py in _initialize_params(self, model, use_logits, input_layer, output_layer)
    264             )
    265         else:
--> 266             raise ValueError("Loss function not recognised.")
    267 
    268         # Define the loss using the loss function

ValueError: Loss function not recognised.
```

I am running it in Google Colab with default environment.

Thanks,
-Amir

**Comment by beat-buesser** [comment]
Hi @amir1m 

The error happens because `KerasClassifier` does not support custom loss functions, only the losses provided by Keras. But for the black-box attack ZOO attack it does not matter what the loss function of the model is because ZOO calculates the loss in the attack independent of the model, but requires for that the models predicts logits. I would recommend to train the model with `model.compile(optimizer='sgd', loss=tf.keras.losses.CategoricalCrossentropy(from_logits=True))`.

You model is still not returning logits with `model.add(Dense(10, activation="sigmoid"))`, it should be `model.add(Dense(10, activation="linear"))`.

**Comment by amir1m** [comment]
Got it! Thanks @beat-buesser ! Worked as expected, 100% ASR. It took ~ 40 minutes to generate 10 MNIST adversarial examples on GPU enabled Colab notebook.

One more thing, how do we enable logging? Looking to find if there is a way to find out how many times ZOO made the queries/call to model to predict while generating an adversarial example? Also, if there are any more black-box evasion attacks implemented or on the ART roadmap. 

Appreciate the help!

**Comment by beat-buesser** [comment]
@amir1m You can just get the root logger with `logger = logging.getLogger()`, `examples/adversarial_training_cifar10.py` shows and example with logging. `HopSkipJump`, `BoundaryAttack`, and `SimBA` are also black-box and available in ART 1.6.

**Comment by amir1m** [comment]
Thanks @beat-buesser ! This is helpful.
